File: src/execution/runtimeTransactionEngine.ts
/**
 * Converted from Python: core/execution/runtime_transaction_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function beginRuntimeTransaction(tick: any = 0, checkpoint_id: any = ""): any {
  var payload: any = py.jsonDumps({"tick": tick, "checkpoint": checkpoint_id}, {sortKeys: true});
  var transaction_id: any = py.slice(py.hashNew("sha256", py.encode(payload, "utf-8")).hexdigest(), null, 32);
  return {"transaction_id": transaction_id, "actions": [], "mutations": [], "transitions": [], "checkpoints": (py.truthy(checkpoint_id) ? [checkpoint_id] : []), "committed": false, "rolled_back": false, "bounded": true};
}
export function commitRuntimeTransaction(transaction: any): any {
  var updated: any = py.pyDict(transaction);
  py.setItem(updated, "committed", true);
  py.setItem(updated, "rolled_back", false);
  return updated;
}
export function rollbackRuntimeTransaction(transaction: any): any {
  var updated: any = py.pyDict(transaction);
  py.setItem(updated, "committed", false);
  py.setItem(updated, "rolled_back", true);
  py.setItem(updated, "actions", []);
  py.setItem(updated, "mutations", []);
  return updated;
}

File: lib/src/execution/runtime_transaction_engine.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

/// Port of core/execution/runtime_transaction_engine.
Map<String, dynamic> beginRuntimeTransaction({
  int tick = 0,
  String checkpointId = '',
}) {
  // Python json.dumps({"tick":tick,"checkpoint":checkpoint_id}, sort_keys=True)
  // default separators (", ", ": "). Keys sorted: checkpoint, tick.
  final String payload =
      '{"checkpoint": ${jsonEncode(checkpointId)}, "tick": $tick}';
  final String transactionId =
      sha256.convert(utf8.encode(payload)).toString().substring(0, 32);

  return <String, dynamic>{
    'transaction_id': transactionId,
    'actions': <dynamic>[],
    'mutations': <dynamic>[],
    'transitions': <dynamic>[],
    'checkpoints':
        checkpointId.isNotEmpty ? <String>[checkpointId] : <String>[],
    'committed': false,
    'rolled_back': false,
    'bounded': true,
  };
}

Map<String, dynamic> commitRuntimeTransaction(
    Map<String, dynamic> transaction) {
  final Map<String, dynamic> updated = <String, dynamic>{...transaction};
  updated['committed'] = true;
  updated['rolled_back'] = false;
  return updated;
}

Map<String, dynamic> rollbackRuntimeTransaction(
    Map<String, dynamic> transaction) {
  final Map<String, dynamic> updated = <String, dynamic>{...transaction};
  updated['committed'] = false;
  updated['rolled_back'] = true;
  updated['actions'] = <dynamic>[];
  updated['mutations'] = <dynamic>[];
  return updated;
}
